[1/N] Nnapi backend delegation preprocess

Amyh11325 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61499**

Added a preprocess function for the delegate to Nnapi backend (internal and external files). 

In the past we had functions and classes for converting to the Nnapi backend. Now, these functions and classes will be wrapped by the delegate API.

### nnapi_backend_preprocess.cpp:

Contains the preprocess function, which uses Pybind to call an existing python function, `convert_model_to_nnapi()`.
- The model is wrapped by a `RecursiveScriptModule`, so that `convert_model_to_nnapi()` can run correctly, since when jumping from Python to C++ to Python, the model loses its original wrapper.
- A tensor, which includes shape, data type, and quantization information, is passed through preprocess's compile_spec to `convert_model_to_nnapi()`.
- Finally, the Nnapi model is serialized for mobile and returned as a string.
### nnapi_backend_lib.cpp:
Contains stub functions for compile and execute, and is necessary for the Nnapi backend to be registered correctly. These will be implemented in a future PR.

**TODO:** implement execute and compile for the delegate API; throw exceptions for incorrect an compile_spec; add OSS tests
**Testing:** Tests were done locally (see D29647123). A simple module was lowered to Nnapi, saved locally, and examined.

Differential Revision: [D29563351](https://our.internmc.facebook.com/intern/diff/D29563351/)

Commit:
[1/N] Nnapi backend delegation preprocess

Added preprocess function for delegation to Nnapi backend (internal and external files).

Differential Revision: [D29563351](https://our.internmc.facebook.com/intern/diff/D29563351/)

[ghstack-poisoned]

iseeyuan commented on torch/csrc/jit/backends/nnapi/nnapi_backend_lib.cpp:
```diff
@@ -0,0 +1,46 @@
+namespace torch {
+namespace jit {
+
+class NnapiBackend : public PyTorchBackendInterface {
```
Add comments to explain why this file with no actual implementation is needed. Add TODO task number T91991928.

raziel replied:
Also, for future PR, this would benefit from some doc about how the overall NNAPI backend works.

Amyh11325 replied:
I'll keep this in mind! Just to clarify, do you mean the NNAPI backend delegation or NNAPI itself?

raziel replied:
Just the backend: things like what it receives as the execution payload, what compile will do with it, and the same for execute.

iseeyuan commented on torch/csrc/jit/backends/nnapi/nnapi_backend_preprocess.cpp:
```diff
@@ -0,0 +1,45 @@
+// Value: a string of the nnapi module, saved for mobile
+c10::IValue preprocess(
+    const torch::jit::Module& mod,
+    const c10::Dict<c10::IValue, c10::IValue>& method_compile_spec,
```
Put an example of method_compile_spec: {"forward": {"inputs": Tensor}} in comment

raziel replied:
I still don't think the current documentation nor example are clear.

Amyh11325 replied:
Could you be a bit more specific? I think that would be really helpful! Maybe I should add more explanations for what input is for? Or clarify that compile_spec is a dictionary within a dictionary?

raziel replied:
Yes, basically the comments I made above.
Also, it would be good if the example shows how "Tensor" is created.

iseeyuan reviewed (approved):
It's not trivial to figure out all the pybind usages and type cases. Great work!

Commit:
Update on "[1/N] Nnapi backend delegation preprocess"


Added a preprocess function for the delegate to Nnapi backend (internal and external files). 

In the past we had functions and classes for converting to the Nnapi backend. Now, these functions and classes will be wrapped by the delegate API.

### nnapi_backend_preprocess.cpp:

Contains the preprocess function, which uses Pybind to call an existing python function, `convert_model_to_nnapi()`.
- The model is wrapped by a `RecursiveScriptModule`, so that `convert_model_to_nnapi()` can run correctly, since when jumping from Python to C++ to Python, the model loses its original wrapper.
- A tensor, which includes shape, data type, and quantization information, is passed through preprocess's compile_spec to `convert_model_to_nnapi()`.
- Finally, the Nnapi model is serialized for mobile and returned as a string.
### nnapi_backend_lib.cpp:
Contains stub functions for compile and execute, and is necessary for the Nnapi backend to be registered correctly. These will be implemented in a future PR.

**TODO:** implement execute and compile for the delegate API; throw exceptions for incorrect an compile_spec; add OSS tests
**Testing:** Tests were done locally (see D29647123). A simple module was lowered to Nnapi, saved locally, and examined.

Differential Revision: [D29563351](https://our.internmc.facebook.com/intern/diff/D29563351/)

[ghstack-poisoned]

raziel commented on torch/csrc/jit/backends/nnapi/nnapi_backend_lib.cpp:
```diff
@@ -0,0 +1,49 @@
+// This file has no implementation yet, but the declarations are necessary to
+// register the backend properly and test preprocess
+// TODO T91991928: implement compile() and execute()
+class NnapiBackend : public PyTorchBackendInterface {
```
Something to address next PR: existing code (mobile/contrib/nnapi/nnapi.h) with different writes "NNApi" rather than "Nnapi" as you did here. I think your way is more principled, but we should be make it consistent (I'd say we refactor the older code with the new form -- @iseeyuan what do you think?).

iseeyuan replied:
I agree to unify them to "Nnapi". I saw "NN", "API", "NNAPI", "nnapi", etc., it would make sense to unify the format in code.

raziel commented on torch/csrc/jit/backends/nnapi/nnapi_backend_preprocess.cpp:
```diff
@@ -0,0 +1,48 @@
+
+namespace py = pybind11;
+
+// Converts model to nnapi and serializes it for mobile
```
nit: at least in docs we should say Android NNAPI.

Most people won't be familiar with what nnapi means.

raziel commented on torch/csrc/jit/backends/nnapi/nnapi_backend_preprocess.cpp:
```diff
@@ -0,0 +1,48 @@
+// Converts model to nnapi and serializes it for mobile
+// Returns a dictionary string with one entry:
+// Key: "NnapiModule"
+// Value: a string of the nnapi module, saved for mobile
```
The value should have a clearer description: i.e. what is this string representation?

raziel commented on torch/csrc/jit/backends/nnapi/nnapi_backend_preprocess.cpp:
```diff
@@ -0,0 +1,48 @@
+// Key: "NnapiModule"
+// Value: a string of the nnapi module, saved for mobile
+//
+// method_compile_spec should contain an input Tensor with the following format:
```
A few questions:
* Does the delegate will only work for "forward" and no other methods? Is there a reason why?
* The value is itself another dictionary, which must have an entry for "inputs" with a value of Tensor type (and shape? and what else?) that matches that of the inputs in the method.

iseeyuan replied:
* I remember @dreiss mentioned that nnapi itself has "forward" as the only entry point. 
* @Amyh11325 is working on a sanity check of the compilation spec. Currently she's trying  to support both a tensor and a list of tensors.

raziel replied:
Sounds good. Just checking, and also worth documenting :-)

raziel commented on torch/csrc/jit/backends/nnapi/nnapi_backend_preprocess.cpp:
```diff
@@ -0,0 +1,48 @@
+      py::module::import("torch.jit._recursive").attr("wrap_cpp_module")(mod);
+  out.attr("eval")();
+  // TODO: throw exception if compile_spec doesn't contain inputs
+  torch::Tensor inp =
```
Per my comment above I think we need to be very clear on what are the requirements of that input tensor.
E.g. dtype, shape, is that enough?

Otherwise users may be confused: e.g. they'll wonder if they need some real data in the tensor, or something else.

raziel reviewed (commented):
Sorry for the late reply.

Please address my comments here, even if the actual code changes go in a separate PR.

Thanks!

Amyh11325 commented:
> Sorry for the late reply.
> 
> Please address my comments here, even if the actual code changes go in a separate PR.
> 
> Thanks!

I incorporated your feedback into this PR! https://github.com/pytorch/pytorch/pull/61752
